autoload/asyncomplete/sources/buffer.vim

```diff
@@ -47,32 +44,34 @@ function! s:should_ignore(opt) abort
     return 0
 endfunction
 
-let s:last_ctx = {}
 function! s:on_event(opt, ctx, event) abort
     if s:should_ignore(a:opt) | return | endif
 
-    if a:event == 'BufWinEnter'
+    if a:event ==# 'BufWinEnter'
         call s:refresh_keywords()
+    elseif a:event ==# 'TextChangedI'
+        let l:typed = a:ctx['typed']
+        if empty(l:typed) " may be a new line, so add the prev line
+            call s:add_line(getline(a:ctx['lnum'] - 1))
+        elseif match(l:typed, '\W$') > -1 " ends with a non-word char, add the typed text
+            call s:add_line(l:typed)
+        endif
```

Comment:
We must catch every char and check it for solving #14 , this is the only way I can think of.

And we add a line only when typed is empty or ends with a non-word char, that means if you type a word, `add_line` function will be invoked only once, so I think there is no perf issues.

what do you think?

Reply:
I created a new commit to reduce unnecessary operations, now I think it can work well.